# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

In [1]:
from scipy.stats import binom

n_tickets = 460
n_seats = 450
p_show = 0.97  # 1 - 3% no-show probability

# X = number of passengers who show up ~ Binomial(n=460, p=0.97)
# "enough seats for everyone" means at most 450 show up
p_enough_seats = binom.cdf(n_seats, n_tickets, p_show)
p_enough_seats

np.float64(0.8844772466215431)

**88.4%** chance of having enough seats for everyone who shows up — overbooking by 10 tickets is a fairly safe bet given the 3% no-show rate.

**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

In [2]:
from scipy.stats import geom

p_success = 0.3

# X = attempt number of the first success ~ Geometric(p=0.3)
# "at least three attempts" means the first two attempts both fail: P(X >= 3)
p_at_least_3 = geom.sf(2, p_success)  # sf(k) = P(X > k) = 1 - cdf(k)
p_at_least_3

np.float64(0.49000000000000005)

**49%** chance the rep needs 3+ attempts — matches the direct formula (1-p)² = 0.7² = 0.49, since "at least 3 attempts" just means the first two both failed.

**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

In [3]:
from scipy.stats import poisson

mean_visits = 500   # average visits per hour
capacity = 550       # server capacity per hour

# X = visits in one hour ~ Poisson(500); overwhelmed if X > 550
p_overwhelmed_hour = poisson.sf(capacity, mean_visits)  # sf(k) = P(X > k)
p_overwhelmed_hour

np.float64(0.01289822084039203)

What is the probability of being overwhelmed at some point during a day? (consider 24hours)

In [4]:
# Treating each of the 24 hours as independent: overwhelmed at some point
# means NOT surviving all 24 hours without being overwhelmed
p_overwhelmed_day = 1 - (1 - p_overwhelmed_hour) ** 24
p_overwhelmed_day

np.float64(0.2677043869515715)

Only **1.3%** chance of being overwhelmed in any single hour, but that compounds to **26.8%** over a full day (24 independent hours) — worth provisioning extra capacity even though the per-hour risk looks small.

**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

In [5]:
from scipy.stats import expon

mean_interarrival = 10  # minutes between customer arrivals

# T = time until next arrival ~ Exponential(mean=10)
p_within_5 = expon.cdf(5, scale=mean_interarrival)
p_within_5

np.float64(0.3934693402873666)

**39.3%** chance the next customer arrives within 5 minutes.

If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

In [6]:
# A break happens when the gap since the last customer exceeds 15 minutes:
# P(T > 15), the memoryless exponential tail
p_break = expon.sf(15, scale=mean_interarrival)
p_break

np.float64(0.22313016014842982)

**22.3%** chance of a 15-minute gap opening up a break — thanks to the exponential distribution's memoryless property, this is just P(T > 15), independent of how long it's already been since the last customer.

**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

In [7]:
from scipy.stats import norm

mean_weight = 150
std_weight = 10

# X ~ Normal(150, 10); P(140 < X < 160)
p_between = norm.cdf(160, mean_weight, std_weight) - norm.cdf(140, mean_weight, std_weight)
p_between

np.float64(0.6826894921370859)

**68.3%** — this is exactly the "68" in the 68-95-99.7 empirical rule, since 140 and 160 are precisely ±1 standard deviation from the mean.

**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

In [8]:
mean_lifetime = 50  # hours

# T ~ Exponential(mean=50); P(T < 30)
p_fail_within_30 = expon.cdf(30, scale=mean_lifetime)
p_fail_within_30

np.float64(0.4511883639059736)

**45.1%** chance of failing within the first 30 hours, out of a 50-hour mean lifetime.